src/net/p2p.rs

```diff
@@ -3613,2219 +3750,4119 @@ impl PeerNetwork {
-        None
-    }
-
-    /// Given an event ID, find the NeighborKey that corresponds to the outbound connection we have
-    /// to the peer the event ID references.  This checks both the conversation referenced by the
-    /// event ID, as well as the reciprocal conversation of the event ID.
-    pub fn find_outbound_neighbor(&self, event_id: usize) -> Option<NeighborKey> {
-        let (is_authenticated, is_outbound, neighbor_key) = match self.peers.get(&event_id) {
-            Some(convo) => (
-                convo.is_authenticated(),
-                convo.is_outbound(),
-                convo.to_neighbor_key(),
-            ),
-            None => {
-                test_debug!("No such neighbor event={}", event_id);
-                return None;
-            }
-        };
-
-        let outbound_neighbor_key = if !is_outbound {
-            let reciprocal_event_id = match self.find_reciprocal_event(event_id) {
-                Some(re) => re,
-                None => {
-                    test_debug!(
-                        "{:?}: no reciprocal conversation for {:?}",
-                        &self.local_peer,
-                        &neighbor_key
-                    );
-                    return None;
-                }
-            };
-
-            let (reciprocal_is_authenticated, reciprocal_is_outbound, reciprocal_neighbor_key) =
-                match self.peers.get(&reciprocal_event_id) {
-                    Some(convo) => (
-                        convo.is_authenticated(),
-                        convo.is_outbound(),
-                        convo.to_neighbor_key(),
-                    ),
-                    None => {
-                        test_debug!(
-                            "{:?}: No reciprocal conversation for {} (event={})",
-                            &self.local_peer,
-                            &neighbor_key,
-                            event_id
+        // Try advancing states until we get blocked.
+        // Once we get blocked, return.
+        loop {
+            let curr_state = self.microblock_tip_sync_state.clone();
+            debug!(
+                "{:?}: Microblock sync state is {:?}",
+                &self.local_peer, &curr_state
+            );
+            match curr_state {
+                MicroblockTipSyncState::CollectPeerURL => {
+                    if self.peers.is_empty() {
+                        debug!(
+                            "{:?}: No peers connected; cannot do microblock tip sync",
+                            &self.local_peer
                         );
-                        return None;
+                        self.microblock_tip_sync_reset();
+                        return (true, faulty_peers, None);
+                    }
+                    // gather URLs from peers
+                    let mut urls = vec![];
+                    for (_event_id, convo) in &self.peers {
+                        if !convo.is_authenticated() || !convo.is_outbound() {
```

Comment:
It's okay to query inbound peers for this -- you can drop the `!convo.is_outbound()`